### PR TITLE
Elevator high level

### PR DESCRIPTION
**Pull request by GaiaZano05**
The classes of the subsystem elevator, inputs, methods and more :D

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import static frc.robot.subsystems.elevator.ElevatorConstants.*;
+
+
+public class Elevator extends SubsystemBase {
+    private static Elevator INSTANCE;
+    private final ElevatorInputs inputs = new ElevatorInputs();
+    private final ElevatorIO io;
+    private final Mechanism2d mechanism2d = new Mechanism2d(MECHANISM_WIDTH, MECHANISM_HEIGHT);
+    private final MechanismRoot2d root = mechanism2d.getRoot("Elevator", ROOT_X, ROOT_Y);
+    private ControlModeValue controlMode;
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: What is `ControlModeValue`?

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Also, why do you need a member of the class if it exists in the inputs?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: Where commands?

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package frc.robot.subsystems.elevator;
+
+import edu.wpi.first.units.Distance;
+import edu.wpi.first.units.MutableMeasure;
+
+public interface ElevatorIO {
+
+    void setPower(double power);
+
+    void getPower();
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: I don't think you need getters in the IO, you have the inputs.

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package frc.robot.subsystems.elevator;
+
+import edu.wpi.first.units.Distance;
+import edu.wpi.first.units.MutableMeasure;
+
+public interface ElevatorIO {
+
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java:8]: Create the instance of the inputs here, then you can have static access from the `Elevator` class, and the IO classes will have it as a member. Auto logged of course. Look at the example subsystem for reference.

File: src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import edu.wpi.first.units.Distance;
+import edu.wpi.first.units.MutableMeasure;
+import org.littletonrobotics.junction.AutoLog;
+
+@AutoLog
+public class ElevatorInputs {
+    public double power;
+    public MutableMeasure<Distance> height;
+    public MutableMeasure<Distance> heightSetpoint;
+    public double acceleration;
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java]: Acceleration should be `MutableMeasure<Velocity<Velocity<Distance>>>`.

**Review by vichik123 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    public MutableMeasure<Distance> getHeight(){
+        return inputs.height;
+    }
+
+    public MutableMeasure<Distance> getHeightSetpoint(){
+        return inputs.heightSetpoint;
+    }
+
+    public void resetEncoder(boolean isBottom) {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Should be a command

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public void setPower(double power) {
+        inputs.power = power;
+        inputs.controlMode = ElevatorIO.ControlMode.PERCENT_OUTPUT;
+    }
+
+    public double getPower() {
+        return inputs.power;
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Is that needed at the subsystem level?

File: src/main/java/frc/robot/subsystems/elevator/ElevatorConstants.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package frc.robot.subsystems.elevator;
+
+public class ElevatorConstants {
+    public static final double MECHANISM_WIDTH = 0;
+    public static final double MECHANISM_HEIGHT = 0;
+    public static final double ROOT_X = 0;
+    public static final double ROOT_Y = 0;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorConstants.java]: What do these mean?

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Robot;
+
+import static frc.robot.subsystems.elevator.ElevatorConstants.*;
+
+
+public class Elevator extends SubsystemBase {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: You don't call `processInputs()`.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 8 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: *Insert Jamaican accent* yeah mon.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: aab075cea2b263cf440a94473ea5c8c88b2e109e

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 2 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: aab075cea2b263cf440a94473ea5c8c88b2e109e

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 3 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java:8]: 58c674daaaa4325d1d3a9a8d3f47ff476fa925d2

File: src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java (hunk 4 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java]: d55db0dc92ba1f0127d5a86a8b7ba058e9af621c

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 5 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: aba4a04d45521e2bf86ed4f24c34aeef1234646b

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 6 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 314e8b65bc79418acae71c0513bf8f82be36a66c

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: That's not what I meant. I mean that the subsystem class should have functions that return commands to be used in command groups.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 5 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Don't forget to implement this in the sim IO when the time comes (Talon FX can be configured to reset automatically when the limit switch is triggered).

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import org.littletonrobotics.junction.Logger;
+
+public class Elevator extends SubsystemBase {
+
+    private static Elevator INSTANCE;
+    private final ElevatorInputs inputs = new ElevatorInputs();
+    private final ElevatorIO io;
+    private final Mechanism2d mechanism2d = new Mechanism2d(MECHANISM_WIDTH, MECHANISM_HEIGHT);
+    private final MechanismRoot2d root = mechanism2d.getRoot("Elevator", ROOT_X, ROOT_Y);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: The mechanism is never updated.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    private ControlModeValue controlMode;
+    private Command lastCommand = null;
+    private Command currentCommand = null;
+    private boolean changedToDefaultCommand = false;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Why are these needed?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        return inputs.height;
+    }
+
+    public MutableMeasure<Distance> getHeightSetpoint() {
+        return inputs.heightSetpoint;
+    }
+
+    public void stop() {
+        inputs.controlMode = null;
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: The control mode should probably not be `null`. What exactly does stop mean? Hold the elevator in place? Just give percent output 0?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import org.littletonrobotics.junction.Logger;
+
+public class Elevator extends SubsystemBase {
+
+    private static Elevator INSTANCE;
+    private final ElevatorInputs inputs = new ElevatorInputs();
+    private final ElevatorIO io;
+    private final Mechanism2d mechanism2d = new Mechanism2d(MECHANISM_WIDTH, MECHANISM_HEIGHT);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: Should be logged with `@AutoLogOutput`.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    public void setPower(double power) {
+        inputs.power = power;
+        inputs.controlMode = ElevatorIO.ControlMode.PERCENT_OUTPUT;
+    }
+
+
+    public void setHeight(MutableMeasure<Distance> height) {
+        inputs.height = height;
+        inputs.controlMode = ElevatorIO.ControlMode.POSITION;
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Where do actual commands go to the IO?

File: src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package frc.robot.subsystems.elevator;
+
+import edu.wpi.first.units.Distance;
+import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Velocity;
+import org.littletonrobotics.junction.AutoLog;
+
+@AutoLog
+public class ElevatorInputs {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java:9]: Should also log limit switches state.

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+import edu.wpi.first.units.MutableMeasure;
+
+public interface ElevatorIO {
+    ElevatorInputs inputs = new ElevatorInputs();
+
+    void setPower(double power);
+
+    void setHeight(MutableMeasure<Distance> height);
+
+    void resetEncoder(boolean isBottom);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: Talon FX can reset its encoder automatically when a limit switch is triggered, so this should not be in the IO, and be implemented only in the sim IO.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+    public void stop() {
+        inputs.controlMode = null;
+    }
+
+    @Override
+    public void periodic() {
+        io.updateInputs(inputs);
+        Logger.processInputs("Elevator", inputs);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Java provides a function that returns this class name: `this.getClass().getSimpleName()`. It should be cached and used here instead of a string -- this is useful if the class' name changes.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 1 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 122e91d55c9bff9c0f0ad7da905917fc680146af

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: 9921811df8729c75faef804d422cc84184d58cfa

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 8 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: 2921fdc3f5ddf12009a0bf12e0b8eab343409f09

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 12 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: cbf18de77bba92907e2683bd70b825d53635c6a8

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 8 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: This is not a relevant commit. Also, you don't need to create `toLog()` and `fromLog()`, `@AutoLog` does it by itself.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 12 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: What I meant is that the mechanism object should be logged as an output of this subsystem. This can be done using the `@AutoLogOutput` annotation.

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 15 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: b267b0f7dd5a16e6ec5933f940d73637ea505cf5

File: src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java (hunk 14 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorInputs.java:9]: eaf49ea3d4d6b5370f8688873f108a0e38f41ad9

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 11 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 751af420830094adca85de67b6942bfd99230177

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 13 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 9921811df8729c75faef804d422cc84184d58cfa

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 16 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 54ac75c8efd87e3df13afc11eef81f56db4a4641

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        return inputs.heightSetpoint;
+    }
+
+    public boolean isBottom(){
+        return inputs.isBottom;
+    }
+
+    public boolean isTop(){
+        return inputs.isTop;
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: What's the usage of these getters?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 13 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: This is good but did you delete the inputs update? Also the name should be `currentHeight`/`setpointHeight` because just "height" is confusing.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 9 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: d72f43b1c29abae9400cba633f56020634a13317

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 10 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: cb724d116e1b626c233b59f3b12084e5a7d42666

File: src/main/java/frc/robot/subsystems/elevator/ElevatorConstants.java (hunk 7 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorConstants.java]: 263c505cbffebc4fd331bcd5570e1bf8f2d7c98c

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 17 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: For the tests, to test if they work/rest encoders?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 8 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: Fixed that, but I cant find the commit

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 13 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 23dec654c9f9bdbecb9b8bcf18041cc462c6c19a

**Review by katzuv — CHANGES_REQUESTED** [review]
Don't forget to run spotless apply

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 9 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: You shouldn't have to log the mechanism to SmartDashboard, provided you log it using AdvantageKit. You do need to update the mechanism angle/height

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: You shouldn't have to log the mechanism to SmartDashboard, provided you log it using AdvantageKit. You do need to update the mechanism angle/height

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 12 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: f8cb901acad6aff5da7da1666edcf023b1d85ed9

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 9 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 35bacb95c36f3e062215937907785abca32fd6ca

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 8 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:17]: No worries, it's cbf18de7. Just so you know, you can find it easily in the Git Log tab in IntelliJ IDEA by filtering on this specific file -- you can read more here: https://www.jetbrains.com/help/idea/log-tab.html#toolbar
Also, it'll be easier to read a comment, fix it, commit it, push the commit, and include the commit hash. That way you won't need to search for the commit later, you'll be able to just copy the hash of the latest commit.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 12 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: I meant logging the `Mechanism2d` object directly. This will make it possible to visualize in AdvantageScope. You can read more [in the documentation](https://github.com/Mechanical-Advantage/AdvantageKit/blob/4c2785d30c61097655b5ad899e200e0c3f8801c3/docs/RECORDING-OUTPUTS.md#mechanism2d).

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 17 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Good point. However, how do you mean to run these tests? If you just want to see the values you can do so using AdvantageScope, so you don't need these getters. If you mean to make some cool command that checks the sensors, the command will still be made in this class, hence you can use the `inputs` object directly in that command method.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 13 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: This is good, however, why did you delete the control mode updates?

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: This commit doesn't create functions that create commands...

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 2 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: This commit has nothing to do with the comment.

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 3 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java:8]: Should be auto logged.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 17 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: ok
3ff7b8aecae42bd4e27314cfb7ba00f869ed13bf

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 13 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 75ae362a55891f69880d4db56c3afe7917acd59b

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 12 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:25]: 7001a6d8aa4eab37110ec957163dec145c658646

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 2 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java]: I referenced the wrong commit, but I did fix it:)

**Review by katzuv — APPROVED** [review]
Good job! 🎼

File: src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java (hunk 3 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/ElevatorIO.java:8]: 08e0a400536838771a47de5a740c3fb4d9ca16a7

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java:1]: 65ec43a10c6159d87a0b2971604dd0a6439a516c

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+import org.littletonrobotics.junction.AutoLogOutput;
+import org.littletonrobotics.junction.Logger;
+
+import static frc.robot.subsystems.elevator.ElevatorConstants.MECHANISM_HEIGHT;
+import static frc.robot.subsystems.elevator.ElevatorConstants.MECHANISM_WIDTH;
+
+public class Elevator extends SubsystemBase {
+
+    private static Elevator INSTANCE;
+    private final ElevatorInputsAutoLogged inputs = new ElevatorInputsAutoLogged();
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: You create this in the `ElevatorIO`, just take it from there.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    public MutableMeasure<Distance> getCurrentHeight() {
+        return inputs.currentHeight;
+    }
+
+    public MutableMeasure<Distance> getHeightSetpoint() {
+        return inputs.heightSetpoint;
+    }
+
+    public Command setHeight(double height){
+        return new RunCommand(()->this.setHeight(height));
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: You can use the `run` function (part of the `SubsystemBase`) instead of the `RunCommand` constructor. This will add requirements as well.

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+
+
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Gah damn.

**Review by vichik123 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 20 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: I will not be referecing a commit, but fixed

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 18 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: 794888d4abe6b3fac0d2a989e56da7c5ea5d4e9a

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java (hunk 19 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/elevator/Elevator.java]: c0366d6f123741f8d53d6540e4f7f57f72e68d71

File: src/main/java/frc/robot/subsystems/elevator/Elevator.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public Command setHeight(double height) {
+        return run(() -> this.setHeight(height));
+    }
+
+    @Override
+    public void periodic() {
+        io.updateInputs(inputs);
+        elevator.setLength(0);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/elevator/Elevator.java]: Shouldn't this be updated according to the height of the elevator?

**Review by vichik123 — CHANGES_REQUESTED** [review]

**Review by vichik123 — APPROVED** [review]
